src/main/java/io/wgj/ctci/Stack.java

```diff
@@ -1,21 +1,80 @@
 package io.wgj.ctci;
 
-public class Stack {
-    int[] stack;
+/**
+ * Stack is a stack data structure backed by a dynamic array. Stack can be
+ * initialized with a starting array size. It's written to develop familiarity
+ * with the stack data structure, and should otherwise not be used.
+ * Maximum allowed array is 1024. If you set a starting capacity larger than
+ * this, the maximum allowed array size is used as the starting capacity. An
+ * UnsupportedOperationException will be thrown when more the maximum allowed
+ * amount of elements are added.
+ */
+public class Stack<E> {
+    static final int MAX_ARRAY_SIZE = 1024;
+    static final int DEFAULT_CAPACITY = 10;
+    Object[] stack;
+    int size;
     int position;
 
-    public Stack(){
-        stack = new int[10];
+    public Stack() {
+        this(DEFAULT_CAPACITY);
+    }
+
+    public Stack(int size) {
+        stack = new Object[size];
+        if (size > MAX_ARRAY_SIZE) {
+            size = MAX_ARRAY_SIZE;
+        }
+        this.size = size;
         position = 0;
     }
 
-    public void push(int i) {
-        stack[position] = i;
+    public void push(E e) {
+        if (size == position) {
+            this.grow();
+        }
+        stack[position] = e;
         position++;
     }
 
-    public int pop(){
+    public E pop() {
         position--;
-        return stack[position];
+        @SuppressWarnings("unchecked")
+        E e = (E) stack[position];
+        return e;
+    }
+
+    private void grow() {
+        int newSize = size << 2;
+        if (newSize > MAX_ARRAY_SIZE) {
+            int leftOver = MAX_ARRAY_SIZE - size;
+            // Squeeze out the remaining space we'd lose.
+            if (leftOver > 0) {
+                newSize = leftOver + size;
+            } else {
+                throw new UnsupportedOperationException();
+            }
+        }
+        Object[] newStack = new Object[newSize];
+        size = newSize;
+        stack = newStack;
+    }
+
+    public E peek() {
+        @SuppressWarnings("unchecked")
+        E e = (E) stack[position - 1];
+        return e;
+    }
+
+    public int size() {
+        return position;
+    }
+
+    public int capacity() {
+        return size;
+    }
+
+    public int maxSize() {
+        return MAX_ARRAY_SIZE;
```

Comment:
Just make MAX_ARRAY_SIZE public, this doesn't need to be a method.  You're using capacity() to do what I'd previously hoped maxSize() would do (which is fine).

Reply:
This is a problem I have with learning Java. I like your suggestion, and even used it for a while, but Sun by way of Oracle makes me think accessing members directly without a getter is worse than killing babies. Could you expand on that?